### PR TITLE
JAPI-329 Utilize TLK during index read

### PR DESCRIPTION
**Pull request by jpmcmu**
- Added partition processor that can use TLK information to filter out partitions
- Changed WSDfuClient code to maintain TLK partitions
- Added partition filtering method to HPCCFile

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [X] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [X] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

**Comment by jpmcmu** [comment]
Original WIP PR: https://github.com/hpcc-systems/hpcc4j/pull/414

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                    {
+                        return sa.compareTo(sb);
+                    }
+                    else
+                    {
+                        return sa.substring(0,comparisonLength).compareTo(sb.substring(0,comparisonLength));
+                    }
+                }
+            default:
+                return 0;
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/CompiledFieldFilter.java:149]: this return value seems misleading, if we can't compare this type, perhaps a exception is appropriate

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/DataPartition.java [hunk 2]
```diff
@@ -415,6 +416,8 @@ public String toString()
                 DataPartition new_dp = new DataPartition(clusterremapper.reviseIPs(dfuparts[i].getCopies()), copyPaths, dfuparts[i].getPartIndex(),
                         dfuparts.length, clusterremapper.revisePort(null), clusterremapper.getUsesSSLConnection(null), filter, fileAccessBlob,
                         fileType);
+                new_dp.isTLK = dfuparts[i].getKind() == DFUFilePartWrapper.FilePartKind.TLK;
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/DataPartition.java:419]: Per offline discussion, istlk() should be used based on this change:
https://github.com/hpcc-systems/hpcc4j/pull/436

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HPCCFile.java [hunk 3]
```diff
@@ -356,7 +360,19 @@ private void createDataParts() throws HpccFileException
                 ClusterRemapper clusterremapper = ClusterRemapper.makeMapper(clusterRemapInfo, fileinfoforread);
                 this.dataParts = DataPartition.createPartitions(fileinfoforread.getFileParts(), clusterremapper,
                         /* maxParts currently ignored anyway */0, filter, fileinfoforread.getFileAccessInfoBlob(), fileType);
+
+                // Check to see if this file has a TLK. The TLK will always be the last partition.
+                // If we do have a TLK remove it from the standard list of data partitions.
+                DataPartition lastPart = this.dataParts[this.dataParts.length-1];
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HPCCFile.java:366]: Minor, do we even care to do this if the file is known not to be index?

**Comment by jpmcmu** [comment]
@rpastrana code review changes made, please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        lastRange.dataPartition = this.dataPartitions[this.dataPartitions.length-1];
+
+        dataPartitionRanges.add(lastRange);
+    }
+
+    public List<DataPartition> findMatchingPartitions(FileFilter filter)
+    {
+        // If we don't have a set of partition ranges. There wasn't a TLK
+        // associated with the file, so return all partitions.
+        if (dataPartitionRanges.size() == 0)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/PartitionProcessor.java:290]: can we bail out if filter == null ?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java [hunk 5]
```diff
@@ -614,8 +614,12 @@ private void readDataInFetch()
             int bytesToRead = 0;
             try
             {
+                // Read at least 512 bytes at a time
+                bytesToRead = Math.max(512,this.dis.available());
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java:618]: what's the significance of 512?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+            wu.setCluster(thorcluster);
+
+            try
+            {
+                HPCCWsWorkUnitsClient client = wsclient.getWsWorkunitsClient();
+                String result = client.createAndRunWUFromECLAndGetResults(wu);
+            }
+            catch(Exception e)
+            {
+                Assert.fail("Test failed with error: " + e.getMessage());
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java:113]: It feels like creating the index file should be externalized, and this test voided if creating the index fails.
I think some of the other test use various org.junit.Assume constructs to do this.

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+                HPCCRecord rangeStart = partitionRangeStart.get(j);
+                
+                String filterStr = null;
+                if (rangeStart.getField(0) instanceof String)
+                {
+                    filterStr = "key = \"" + rangeStart.getField(0).toString() + "\"";
+                }
+                else
+                {
+                    filterStr = "key = " + rangeStart.getField(0).toString();
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java:137]: I didn't notice a write up on how one would compose the filters

**Review by rpastrana — COMMENTED** [review]
@jpmcmu a few comments

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java (hunk 5 above)
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/RowServiceInputStream.java:618]: It was a value I found through benchmarking. I will add a constant for it and add a comment explaining it usage

**Comment by jpmcmu** [comment]
@rpastrana please review

**Review by rpastrana — APPROVED** [review]
@jpmcmu looks good

**Pull request merged**